### ISSUE TITLE
How can I remove objectness loss from training?

### ISSUE THREAD

**Issue by imyhxy** [ultralytics/yolov5#2010]
I'm doing a experiment on coco128 dataset and try to remove the `objectness` term from loss. Now I just simply remove the caculation of `lobj` and use a tensor with shape of `[bs, 3, h, w, 80]` to calculate the `lcls`, but the loss doesn't seem to coverage anymore. Should I use other loss functions after removing objectness? For example, `softmax` or `QFocalLoss`. Or this idea just doesn't make sense to `yolov5`.

```python
class ComputeNoObjectnessLoss(ComputeLoss):
    def __call__(self, p, gts):
        device = gts.device
        lcls, lbox = torch.zeros(1, device=device), torch.zeros(1, device=device)
        tcls, tbox, indices, anchors = self.build_targets(p, gts)  # targets

        for i, pi in enumerate(p):  # layer index, layer predictions
            b, a, gj, gi = indices[i]  # image, anchor, gridy, gridx
            tcls = torch.zeros_like(pi[..., 4:], device=device)  # target obj

            n = b.shape[0]  # number of targets
            if n:
                ps = pi[b, a, gj, gi]  # prediction subset corresponding to targets

                # Regression
                pxy = ps[:, :2].sigmoid() * 2. - 0.5
                pwh = (ps[:, 2:4].sigmoid() * 2) ** 2 * anchors[i]
                pbox = torch.cat((pxy, pwh), 1)  # predicted box
                iou = bbox_iou(pbox.T, tbox[i], x1y1x2y2=False, CIoU=True)  # iou(prediction, target)
                lbox += (1.0 - iou).mean()  # iou loss

                # Objectness
                # tobj[b, a, gj, gi] = ((1.0 - self.gr) + self.gr * iou.detach().clamp(0).type(tobj.dtype)).unsqueeze(-1)

                # Classification
                tcls[b, a, gj, gi, tcls[i]] = self.cp
            clsi = self.BCEcls(pi[..., 4:], tcls)
            lcls += clsi * self.balance[i]  # BCE
            if self.autobalance:
                self.balance[i] = self.balance[i] * 0.9999 + 0.0001 / clsi.detach().item()

        if self.autobalance:
            self.balance = [x / self.balance[self.ssi] for x in self.balance]
        lbox *= self.hyp['box']
        lcls *= self.hyp['cls']
        bs = p[0].shape[0]  # batch size

        loss = lbox + lcls
        return loss * bs, torch.cat((lbox, torch.tensor([0.], device=device), lcls, loss)).detach()

```

**Comment by glenn-jocher** [comment]
@imyhxy you can set any loss component's hyperparameter gain to zero if you want to remove it's effect, though naturally training will no longer work if you do this:
https://github.com/ultralytics/yolov5/blob/046c37e465d60b9047ff0e91f44594817e5a935c/data/hyp.scratch.yaml#L16

**Comment by imyhxy** [comment]
@glenn-jocher Now I multiply the objectness and the classification label, and then only calculate the classification loss. I still haven't got the final result, but it seems to work, thank you.

**Comment by imyhxy** [comment]
@glenn-jocher Got `37.0AP` with `yolov5s`, but GPU memory usage is severely unbalanced, GPU-0 reach `14G`, and other GPU is only about `3G`.

**Comment by glenn-jocher** [comment]
@imyhxy hi, I don't understand. Did you train with your modified loss function and obtained 37 AP with YOLOv5s?

If GPU memory is low you may want to increase batch size or image size to exploit that additional memory.

**Comment by imyhxy** [comment]
@glenn-jocher Yes, I trained with the custom loss on the latest YOLO code (9 days ago), and the `val` result is:
![image](https://user-images.githubusercontent.com/12193450/106377578-f4700580-63d8-11eb-940d-385f31fda428.png)

I don't think I can increase the `batch size` anymore because the GPU memory usage looks like below, the GPU-0 is almost reach the maximum:
![image](https://user-images.githubusercontent.com/12193450/106377806-96dcb880-63da-11eb-803f-c39be72c1eb6.png)

**Comment by glenn-jocher** [comment]
@imyhxy well then it looks like your custom loss function is a success (!), as it's improving upon the baseline results (36.8).

If you are using a high number of GPUs (i.e. 8 as in your case, its possible that testing may incur higher GPU usage on device 0 due to the way the arguments are passed to test.py. This has been fixed in an R&D branch but not in master. This is because the subdivided batch size is used for training but the full batch size is passed to test.py when testing, causing the imbalance you are seeing above. You can manually adjust the test.py and testloader --batch-size arguments to compensate as an interim solution.

In any case though, be aware that increasing batch-size past 64 may slow down loss convergence as there are less optimization updates during training. (It logically follows that as batch-size trends to infinity the optimization step size will also increase to infinity, inherently degrading training.) Where that point is in the batch-size space exactly, I do not know.
